### ISSUE TITLE
Header inclusion problem when building plugins with Homebrew installs on MacOS

### ISSUE THREAD

**Issue by ckreibich** [zeek/zeek#1468]
@philrz reported corelight/zeek-community-id#15, in which Zeek header inclusion dies a few levels into ` zeek/plugin/Plugin.h` when building on the Github workflows MacOS VM:

```
In file included from /Users/runner/.zkg/clones/package/zeek-community-id/src/Plugin.cc:2:
In file included from /Users/runner/.zkg/clones/package/zeek-community-id/src/Plugin.h:3:
In file included from /usr/local/include/zeek/plugin/Plugin.h:11:
In file included from /usr/local/include/zeek/logging/WriterBackend.h:7:
In file included from /usr/local/include/zeek/threading/MsgThread.h:5:
In file included from /usr/local/include/zeek/threading/Queue.h:10:
In file included from /usr/local/include/zeek/Reporter.h:12:
In file included from /Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_set:364:
In file included from /Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:18:
/Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:317:9: error: no member named 'signbit' in the global namespace
using ::signbit;
      ~~^
/Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:318:9: error: no member named 'fpclassify' in the global namespace
using ::fpclassify;
      ~~^
/Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:319:9: error: no member named 'isfinite' in the global namespace; did you mean 'finite'?
using ::isfinite;
      ~~^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/math.h:749:12: note: 'finite' declared here
extern int finite(double)
           ^
```
This problem isn't unique to us. Quick Googling suggests lots of projects have hit this, and there's a good explanation [here](https://stackoverflow.com/questions/58628377/catalina-c-using-cmath-headers-yield-error-no-member-named-signbit-in-th).

It happens only with the Homebrew-based install and not when installing Zeek from source. I've tracked this down to a path discrepancy in the `zeek-config --include_dir` directories: Homebrew wants to use a MacOSX 10.15 SDK (which fails) whereas the local build selects the 11.1 one installed with XCode (which succeeds).

A really crude workaround is to swap out their include paths in `zeek-config` (see again corelight/zeek-community-id#15 for details).

It seems clear that this isn't a specific Zeek package-level problem, so I'm wondering whether there's a Zeek-level fix. Maybe our `MacDependencyPaths` CMake module needs a tweak, or the Homebrew formula needs fixing. Or the 10.15 SDK is just busted in this regard? I'll leave this to the Mac folks in the audience. #1368 looks related, too.

This had me wondering whether we should allow `zeek-config` to report modified paths post-installation. Seems that occasionally we hit external circumstances in which this would be handy. Here we could have prefixed the working include path (though that'd clearly still be a hack), and a similar ability to override paths would have helped with a FreeBSD staged install (see [mailing list discussion](https://lists.zeek.org/archives/list/zeek-dev@lists.zeek.org/thread/5QU22ZTRVHJRKGIWNL3KF6ICV3VY4S43/)).

**Comment by jsiwek** [comment]
If the Homebrew Bottle's `zeek-config` has already baked in the SDK path from the host system doing the bottling, my hunch is to call that the problem.  Not sure I looked in the right place for bottles (https://bintray.com/homebrew/bottles/zeek#files), but seems to be the case, so I wonder if there's some extra path substitution that needs to be done during the (un)bottling process (I'm not familiar with what's typical/possible).

It could be possible for Zeek to alter the CMake logic that produces `zeek-config` and omit include paths that are based on `CMAKE_OSX_SYSROOT`, but not sure that's best for a couple reasons:
* Are we supposed to be in the business of knowing every possible path that's "off limits" per OS/packaging-system ?
* Think it does need to report those paths for correct behavior -- if a plugin includes a Zeek header that in turn includes an external dependency header, `zeek-config` really should report a path to where that can be found, wherever it is.

**Comment by ckreibich** [comment]
> If the Homebrew Bottle's zeek-config has already baked in the SDK path from the host system doing the bottling, my hunch is to call that the problem.

I think that's right Jon — I just simply substituted out the SDK path, making the full return value of `zeek-config --include_dir` this:
```
/usr/local/Cellar/zeek/4.0.0/include:/usr/local/Cellar/zeek/4.0.0/include/zeek:/usr/include:/usr/include:/usr/local/opt/openssl@1.1/include::
```
... it worked. The middle `/usr/include` ones would normally have been `/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include`.

**Comment by bbannier** [comment]
Just watching this from the side (and we also bakethe sysroot into the include paths returned by Spicy's `spicy-config`), but I wonder if baking in the sysroot is at all a good idea. The compiler which the user would invoke from that output should already know its sysroot (which might deviate from the one used to compile and configure the tool here), and I wonder whether the config query tool should just not include it.

**Comment by philrz** [comment]
@bbannier: I'm a novice at the Zeek-specific areas of this, but I was hacking at this one over the weekend and found myself wondering something similar. @ckreibich mentioned that Mac users in the Zeek Dev team tend to favor MacPorts over Homebrew, so I tried installing that for the first time and I found that the Community ID package had no problem installing atop it, and in that case I saw no Mac-specific entries in the `include_dir` lines in its `zeek-config`. I didn't know if that meant the MacPorts one was now ill-equipped for handling _other_ packages, but suffice to say, it made me wonder.

**Comment by bbannier** [comment]
Funnily enough, these sysroot paths seem to come out of `PCAP_INCLUDE_DIR` and `ZLIB_INCLUDE_DIR` [here](https://github.com/zeek/zeek/blob/ea8367713b93bdf7720b83e5c5f25644e30e038d/zeek-config.in#L34-L35) which indeed seem to contain the headers the `FindXYZ.cmake` scripts are looking for (`pcap.h` and `zlib.h`).

Looking at the Homebrew formula for Zeek I see that it [uses the system pcap](https://github.com/Homebrew/homebrew-core/blob/0e4444cbc26081d196a216cb5f574bf6d9196cdf/Formula/zeek.rb#L29) (the zlib dependency isn't declared), while macports uses its own packages which are probably not in these sysroot paths (haven't checked).

**Comment by bbannier** [comment]
Fixing the Homebrew formula to more closely mirror what macports does seems to lead to the sysroots going away,

```.diff
diff --git a/Formula/zeek.rb b/Formula/zeek.rb
index 892a039de8..37251717cf 100644
--- a/Formula/zeek.rb
+++ b/Formula/zeek.rb
@@ -24,9 +24,10 @@ class Zeek < Formula
   depends_on macos: :mojave
   depends_on "openssl@1.1"
   depends_on "python@3.9"
+  depends_on "libpcap"
+  depends_on "zlib"

   uses_from_macos "flex"
-  uses_from_macos "libpcap"

   resource "pcap-test" do
     url "https://raw.githubusercontent.com/zeek/zeek/59ed5c75f190d4401d30172b9297b3592dd72acf/testing/btest/Traces/http/get.trace"
@@ -44,7 +45,9 @@ class Zeek < Formula
                       "-DCAF_ROOT=#{Formula["caf"].opt_prefix}",
                       "-DOPENSSL_ROOT_DIR=#{Formula["openssl@1.1"].opt_prefix}",
                       "-DZEEK_ETC_INSTALL_DIR=#{etc}",
-                      "-DZEEK_LOCAL_STATE_DIR=#{var}"
+                      "-DZEEK_LOCAL_STATE_DIR=#{var}",
+                      "-DPCAP_ROOT_DIR=#{Formula["libpcap"].opt_prefix}",
+                      "-DZLIB_ROOT=#{Formula["zlib"].opt_prefix}"
       system "make", "install"
     end
   end
```

```
$ zeek-config --include_dir
/usr/local/Cellar/zeek/4.0.0_3/include:/usr/local/Cellar/zeek/4.0.0_3/include/zeek:/usr/local/opt/libpcap/include:/usr/local/opt/zlib/include:/usr/local/opt/openssl@1.1/include::
```

**Comment by philrz** [comment]
Thanks @bbannier! That helps a lot. I set up a scratch test environment with your proposed changes and confirmed that it has the intended effect of the Community ID package now installing via `zkg` without complaint. I also went through the [Homebrew PR prep steps](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request) to confirm that it passes all their requirements, so I could easily turn my the [branch-in-a-fork](https://github.com/Homebrew/homebrew-core/compare/master...brimdata:zeek-avoid-sysroots) into a PR. I've had good luck getting a couple of my other recent minor improvements merged to Homebrew's Zeek formula (Homebrew/homebrew-core#74498, Homebrew/homebrew-core#74518) so I imagine this one would be well-received as well. I figured I'd just check one last time over here to see if there's consensus that this seems like solid incremental improvement. It certainly seems that way to _me_ because it would allow me to declare victory on what I originally floated via corelight/zeek-community-id#15, but I have less of a sense of whether it could negatively impact anything else in other Zeek use cases. Since it makes its config more like MacPorts as you say, and @ckreibich indicated MacPorts is favored by the Mac users within the Zeek Dev team, I'm guessing the change is non-controversial. I know you folks don't claim to speak for the Homebrew install, and neither do I, but it's definitely "out there" and Mac users stumble onto it and see it as "Zeek", so it seems worth the improvement if nobody can think of a reason not to.

I'll wait to hear if you or anyone else voices concerned before putting up a PR. Thanks again!
